File: server/events.py
"""
server/events.py — typed WS event dataclasses.

Single source of truth for all WebSocket message shapes.
Replace the ad-hoc dict construction in _translate_event()
with these structured types.

The frontend mirrors these types in web/src/types/events.ts.
"""

from __future__ import annotations

from dataclasses import dataclass, field, asdict
from typing import Any, Literal


def _to_dict(obj) -> dict:
    """Serialize a dataclass to dict.

    Only skip None values and the 'type' discriminator is always included.
    Empty strings and empty containers are preserved — the frontend may
    depend on their presence (e.g. ev.error || 'default').
    """
    result = {}
    for k, v in asdict(obj).items():
        if v is None:
            continue
        result[k] = v
    return result


# ── Status events ─────────────────────────────────────────────────────


@dataclass
class WsStatus:
    type: Literal["status"] = "status"
    status: str = ""            # running | completed | failed | finish | gave_up | compacted
    message: str = ""
    error: str = ""
    result: dict | None = None  # {summary, steps_taken, total_tokens}
    timestamp: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


# ── Thought ───────────────────────────────────────────────────────────


@dataclass
class WsThought:
    type: Literal["thought"] = "thought"
    content: str = ""
    step: int = 0
    child_session_id: str = ""
    timestamp: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


@dataclass
class WsThoughtDelta:
    """Streaming thought token — pushed in real-time during LLM generation."""
    type: Literal["thought_delta"] = "thought_delta"
    text: str = ""
    step: int = 0
    child_session_id: str = ""
    timestamp: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


# ── Tool call ─────────────────────────────────────────────────────────


@dataclass
class WsToolCall:
    type: Literal["tool_call"] = "tool_call"
    name: str = ""
    params: dict = field(default_factory=dict)
    step: int = 0
    id: str = ""
    child_session_id: str = ""
    timestamp: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


# ── Observation ───────────────────────────────────────────────────────


@dataclass
class WsObservation:
    type: Literal["observation"] = "observation"
    tool_name: str = ""
    output: str = ""
    error: str = ""
    status: str = ""
    step: int = 0
    id: str = ""
    diff: str = ""
    child_session_id: str = ""
    timestamp: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


# ── Reflection ────────────────────────────────────────────────────────


@dataclass
class WsReflection:
    type: Literal["reflection"] = "reflection"
    content: str = ""
    timestamp: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


# ── Subagent events ───────────────────────────────────────────────────


@dataclass
class WsSubagentStart:
    type: Literal["subagent_start"] = "subagent_start"
    child_session_id: str = ""
    agent_name: str = ""
    timestamp: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


@dataclass
class WsSubagentStop:
    type: Literal["subagent_stop"] = "subagent_stop"
    child_session_id: str = ""
    status: str = ""
    timestamp: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


# ── Approval events ───────────────────────────────────────────────────


@dataclass
class WsApprovalRequired:
    type: Literal["approval_required"] = "approval_required"
    request_id: str = ""
    tool_name: str = ""
    params: dict = field(default_factory=dict)
    thought: str = ""
    decision_reason: str = ""
    tool_use_id: str = ""
    permission_mode: str = ""
    risk_level: str = ""
    timestamp: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


@dataclass
class WsApprovalTimeout:
    type: Literal["approval_timeout"] = "approval_timeout"
    request_id: str = ""
    timestamp: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


# ── Plan ready ────────────────────────────────────────────────────────


@dataclass
class WsPlanReady:
    type: Literal["plan_ready"] = "plan_ready"
    plan_text: str = ""
    contract: dict | None = None
    revision: int = 0
    max_revisions: int = 5
    result: dict | None = None
    timestamp: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


# ── Worktree resolved ─────────────────────────────────────────────────


@dataclass
class WsWorktreeResolved:
    type: Literal["worktree_resolved"] = "worktree_resolved"
    child_session_id: str = ""
    action: str = ""
    status: str = ""
    message: str = ""
    timestamp: str = ""

    def to_dict(self) -> dict:
        return _to_dict(self)


# ── Discriminated union ───────────────────────────────────────────────

WsEvent = (
    WsStatus | WsThought | WsThoughtDelta | WsToolCall | WsObservation | WsReflection
    | WsSubagentStart | WsSubagentStop
    | WsApprovalRequired | WsApprovalTimeout | WsPlanReady
    | WsWorktreeResolved
)

File: web/src/types/events.ts
/**
 * Typed WebSocket event discriminated union.
 *
 * Mirrors server/events.py dataclass shapes.  The 'type' field is the
 * discriminator — handler maps use Extract<> for type narrowing.
 *
 * Pattern: CorvidAgent shared type layer + react-socket typed events.
 * Source: https://github.com/CorvidLabs/corvid-agent/issues/957
 */

// ── Status ──────────────────────────────────────────────────────────────

export interface WsStatusEvent {
  type: "status";
  status: "running" | "completed" | "failed" | "finish" | "gave_up" | "cancelled" | "compacted";
  message?: string;
  error?: string;
  result?: { summary?: string; steps_taken?: number; total_tokens?: number };
  timestamp?: string;
  step?: number;
  duration_ms?: number;
  token_estimate?: number;
  child_session_id?: string;
}

// ── Thought / Reflection ────────────────────────────────────────────────

export interface WsThoughtEvent {
  type: "thought";
  content: string;
  timestamp?: string;
  step?: number;
  duration_ms?: number;
  token_estimate?: number;
  child_session_id?: string;
}

export interface WsThoughtDeltaEvent {
  type: "thought_delta";
  text: string;
  timestamp?: string;
  step?: number;
  child_session_id?: string;
}

export interface WsReflectionEvent {
  type: "reflection";
  content: string;
  timestamp?: string;
  step?: number;
  duration_ms?: number;
  token_estimate?: number;
}

// ── Tool call / Observation ─────────────────────────────────────────────

export interface WsToolCallEvent {
  type: "tool_call";
  name: string;
  params?: Record<string, unknown>;
  id?: string;
  timestamp?: string;
  step?: number;
  duration_ms?: number;
  token_estimate?: number;
  child_session_id?: string;
}

export interface WsObservationEvent {
  type: "observation";
  tool_name?: string;
  output?: string;
  error?: string;
  status?: string;
  id?: string;
  paired?: boolean;
  diff?: string;
  timestamp?: string;
  step?: number;
  duration_ms?: number;
  token_estimate?: number;
  child_session_id?: string;
}

// ── Subagent ────────────────────────────────────────────────────────────

export interface WsSubagentStartEvent {
  type: "subagent_start";
  child_session_id: string;
  agent_name?: string;
  timestamp?: string;
  step?: number;
}

export interface WsSubagentStopEvent {
  type: "subagent_stop";
  child_session_id: string;
  status?: string;
  timestamp?: string;
  step?: number;
}

// ── Approval ────────────────────────────────────────────────────────────

export interface WsApprovalRequiredEvent {
  type: "approval_required";
  request_id: string;
  tool_name: string;
  params?: Record<string, unknown>;
  thought?: string;
  decision_reason?: string;
  tool_use_id?: string;
  permission_mode?: string;
  risk_level?: string;
  timestamp?: string;
  step?: number;
}

export interface WsApprovalTimeoutEvent {
  type: "approval_timeout";
  request_id: string;
  timestamp?: string;
}

// ── Plan ────────────────────────────────────────────────────────────────

export interface WsPlanReadyEvent {
  type: "plan_ready";
  plan_text?: string;
  contract?: Record<string, unknown> | null;
  revision?: number;
  max_revisions?: number;
  result?: { summary?: string; steps_taken?: number; total_tokens?: number };
  timestamp?: string;
  step?: number;
}

// ── Worktree ────────────────────────────────────────────────────────────

export interface WsWorktreeResolvedEvent {
  type: "worktree_resolved";
  child_session_id: string;
  action: string;
  status: string;
  message?: string;
  timestamp?: string;
  step?: number;
}

// ── Memory activity ─────────────────────────────────────────────────────

export interface WsMemoryRecallEvent {
  type: "memory_recall";
  injected_count: number;
  candidate_count: number;
  omitted_count: number;
  top_names: string[];
  timestamp?: string;
}

export interface WsMemoryWrittenEvent {
  type: "memory_written";
  name: string;
  description: string;
  source: string;
  confidence: number;
  timestamp?: string;
}

// ── Discriminated union ─────────────────────────────────────────────────

export type WsMessage =
  | WsStatusEvent
  | WsThoughtEvent
  | WsThoughtDeltaEvent
  | WsReflectionEvent
  | WsToolCallEvent
  | WsObservationEvent
  | WsSubagentStartEvent
  | WsSubagentStopEvent
  | WsApprovalRequiredEvent
  | WsApprovalTimeoutEvent
  | WsPlanReadyEvent
  | WsWorktreeResolvedEvent
  | WsMemoryRecallEvent
  | WsMemoryWrittenEvent;

// ── Typed handler utility ───────────────────────────────────────────────

/** Narrow a WsMessage to a specific subtype. */
export type WsMessageOfType<T extends WsMessage["type"]> = Extract<WsMessage, { type: T }>;

// ── Transport-level envelope ────────────────────────────────────────────

/**
 * WsMessage with optional trace sequence number injected by the backend
 * transport layer (EventBus WebSocket broadcast + /timeline REST).
 *
 * ``seq`` is NOT a domain property of individual event types — it is
 * a per-session monotonic counter that the backend stamps on every
 * event at persist/broadcast time.  Use this type in store handlers
 * and transport adapters; render components should use ``WsMessage``.
 */
export type WsEnvelope = WsMessage & { seq?: number };
